File: api/fulfillment/zoho-billing.js
const axios = require('axios');

// In-memory cache with timestamp (works per serverless instance)
let cachedAccessToken = null;
let tokenExpiration = null;
let tokenRefreshPromise = null;

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

async function refreshBillingAccessToken() {
    // Use ZOHO_BILLING_* credentials (separate from portal credentials)
    const refreshToken = process.env.ZOHO_BILLING_REFRESH_TOKEN;
    const clientId = process.env.ZOHO_BILLING_CLIENT_ID;
    const clientSecret = process.env.ZOHO_BILLING_CLIENT_SECRET;

    if (!refreshToken || !clientId || !clientSecret) {
        throw new Error('Missing ZOHO_BILLING_* credentials in environment variables');
    }

    const response = await axios.post('https://accounts.zoho.com/oauth/v2/token', null, {
        params: {
            refresh_token: refreshToken,
            client_id: clientId,
            client_secret: clientSecret,
            grant_type: 'refresh_token',
        },
    });

    return response.data;
}

// Get Zoho Billing access token with in-memory caching
async function getZohoBillingAccessToken() {
    const currentTime = Date.now();

    // Return cached token if still valid
    if (cachedAccessToken && tokenExpiration && currentTime < tokenExpiration) {
        return cachedAccessToken;
    }

    try {
        if (tokenRefreshPromise) {
            return await tokenRefreshPromise;
        }

        tokenRefreshPromise = (async () => {
            let attempt = 0;
            while (attempt < 3) {
                try {
                    const data = await refreshBillingAccessToken();
                    cachedAccessToken = data.access_token;
                    const expiresIn = data.expires_in || 3600;
                    tokenExpiration = Date.now() + (expiresIn - 60) * 1000; // Subtract 60s buffer
                    return cachedAccessToken;
                } catch (error) {
                    const errorData = error.response?.data;
                    const message = errorData?.error_description || error.message || '';
                    const isRateLimit = message.toLowerCase().includes('too many requests');
                    attempt += 1;
                    if (!isRateLimit || attempt >= 3) {
                        throw error;
                    }
                    await sleep(1000 * attempt);
                }
            }
            throw new Error('Failed to refresh Zoho Billing access token');
        })();

        return await tokenRefreshPromise;
    } catch (error) {
        console.error('Error obtaining Zoho Billing access token:', error.response?.data || error.message);
        throw new Error('Failed to get Zoho Billing access token');
    } finally {
        tokenRefreshPromise = null;
    }
}

// Get custom field value from invoice
function getCustomFieldValue(invoice, fieldLabel) {
    const field = invoice.custom_fields?.find(cf => cf.label === fieldLabel);
    return field?.value || '';
}

// Helper function to add delay (rate limiting)
function delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Fetch subscriptions in batches to avoid rate limiting
// Zoho API limit: 30 requests per minute threshold
// We use 3 per batch with 7 second delay = ~25 requests per minute (safe margin)
async function fetchSubscriptionDetailsInBatches(subscriptions, batchSize = 3, delayMs = 7000) {
    const results = [];

    console.log(`Fetching details for ${subscriptions.length} subscriptions in batches of ${batchSize}...`);

    for (let i = 0; i < subscriptions.length; i += batchSize) {
        const batch = subscriptions.slice(i, i + batchSize);
        console.log(`Processing batch ${Math.floor(i / batchSize) + 1} of ${Math.ceil(subscriptions.length / batchSize)}...`);

        const batchPromises = batch.map(sub =>
            getOrderDetails(sub.subscription_id).catch(err => {
                console.error(`Failed to fetch details for ${sub.subscription_id}:`, err.message);
                return sub; // Return minimal data if fetch fails
            })
        );

        const batchResults = await Promise.all(batchPromises);
        results.push(...batchResults);

        // Add delay between batches to avoid rate limiting (30 req/min threshold)
        if (i + batchSize < subscriptions.length) {
            await delay(delayMs);
        }
    }

    console.log(`Completed fetching ${results.length} subscription details`);
    return results;
}

// Get pending orders (both invoices and subscriptions with cf_shipping_status=New Manual Order)
async function getPendingOrders() {
    const accessToken = await getZohoBillingAccessToken();
    const orgId = process.env.ZOHO_ORG_ID;

    try {
        // Fetch subscriptions (1 API call)
        const subscriptionsResponse = await axios.get(`https://www.zohoapis.com/billing/v1/subscriptions`, {
            headers: {
                'Authorization': `Zoho-oauthtoken ${accessToken}`,
                'X-com-zoho-subscriptions-organizationid': orgId
            },
            params: { per_page: 200 }
        });

        const subscriptions = subscriptionsResponse.data.subscriptions || [];

        // Filter subscriptions with status "New Manual Order"
        const filteredSubscriptions = subscriptions.filter(subscription => {
            const status = (subscription.cf_shipping_status || '').trim().toLowerCase();
            return status === 'new manual order';
        });

        console.log(`Found ${filteredSubscriptions.length} subscriptions with status "New Manual Order"`);

        // Return subscriptions with custom field addresses
        return filteredSubscriptions.map(sub => ({
            ...sub,
            _source: 'subscription'
        }));
    } catch (error) {
        console.error('Error fetching pending orders:', error.response?.data || error.message);
        throw error;
    }
}

// Get shipped orders (subscriptions with customer addresses)
async function getShippedOrders() {
    const accessToken = await getZohoBillingAccessToken();
    const orgId = process.env.ZOHO_ORG_ID;

    try {
        // Fetch subscriptions (1 API call)
        const subscriptionsResponse = await axios.get(`https://www.zohoapis.com/billing/v1/subscriptions`, {
            headers: {
                'Authorization': `Zoho-oauthtoken ${accessToken}`,
                'X-com-zoho-subscriptions-organizationid': orgId
            },
            params: { per_page: 200 }
        });

        const subscriptions = subscriptionsResponse.data.subscriptions || [];

        // Filter subscriptions with status "Shipped"
        const filteredSubscriptions = subscriptions.filter(subscription => {
            return subscription.cf_shipping_status === 'Shipped';
        });

        console.log(`Found ${filteredSubscriptions.length} subscriptions with status "Shipped"`);

        // Return subscriptions with custom field addresses
        return filteredSubscriptions.map(sub => ({
            ...sub,
            _source: 'subscription'
        }));
    } catch (error) {
        console.error('Error fetching shipped orders:', error.response?.data || error.message);
        throw error;
    }
}

// Get subscriptions pending T-Mobile activation
async function getPendingTmobileActivations() {
    const accessToken = await getZohoBillingAccessToken();
    const orgId = process.env.ZOHO_ORG_ID;

    try {
        const subscriptionsResponse = await axios.get(`https://www.zohoapis.com/billing/v1/subscriptions`, {
            headers: {
                'Authorization': `Zoho-oauthtoken ${accessToken}`,
                'X-com-zoho-subscriptions-organizationid': orgId
            },
            params: { per_page: 200 }
        });

        const subscriptions = subscriptionsResponse.data.subscriptions || [];

        const filteredSubscriptions = subscriptions.filter((subscription) => {
            const simCard = (subscription.cf_sim_card_number || '').trim();
            const activeOnTmobile = (subscription.cf_active_on_tmobile || '').trim().toLowerCase();
            const status = (subscription.status || subscription.subscription_status || '').trim().toLowerCase();

            return simCard !== '' && activeOnTmobile === 'no' && status === 'live';
        });

        console.log(`Found ${filteredSubscriptions.length} subscriptions pending T-Mobile activation`);

        // Fetch full details for each subscription to get all SIM card fields
        // Use Promise.all for parallel requests (faster than sequential)
        const detailedSubscriptions = await Promise.all(
            filteredSubscriptions.map(async (sub) => {
                try {
                    const detailResponse = await axios.get(
                        `https://www.zohoapis.com/billing/v1/subscriptions/${sub.subscription_id}`,
                        {
                            headers: {
                                'Authorization': `Zoho-oauthtoken ${accessToken}`,
                                'X-com-zoho-subscriptions-organizationid': orgId
                            }
                        }
                    );
                    const fullSub = detailResponse.data.subscription;
                    // Extract custom field values from the custom_fields array
                    const customFieldValues = {};
                    if (fullSub.custom_fields) {
                        fullSub.custom_fields.forEach(field => {
                            if (field.api_name) {
                                customFieldValues[field.api_name] = field.value;
                            }
                        });
                    }
                    // Merge: keep original list data, add custom field values
                    return {
                        ...sub,
                        ...customFieldValues,
                        _source: 'subscription'
                    };
                } catch (err) {
                    console.error(`Failed to fetch details for subscription ${sub.subscription_id}:`, err.message);
                    return { ...sub, _source: 'subscription' };
                }
            })
        );

        return detailedSubscriptions;
    } catch (error) {
        console.error('Error fetching pending T-Mobile activations:', error.response?.data || error.message);
        throw error;
    }
}

// Get order details by invoice ID
async function getOrderDetails(invoiceId) {
    const accessToken = await getZohoBillingAccessToken();
    const orgId = process.env.ZOHO_ORG_ID;

    try {
        const response = await axios.get(`https://www.zohoapis.com/billing/v1/invoices/${invoiceId}`, {
            headers: {
                'Authorization': `Zoho-oauthtoken ${accessToken}`,
                'X-com-zoho-subscriptions-organizationid': orgId
            }
        });

        return response.data.invoice;
    } catch (error) {
        console.error('Error fetching invoice details:', error.response?.data || error.message);
        throw error;
    }
}

// Update invoice custom fields
async function updateInvoiceFields(invoiceId, customFields) {
    const accessToken = await getZohoBillingAccessToken();
    const orgId = process.env.ZOHO_ORG_ID;

    try {
        const response = await axios.put(
            `https://www.zohoapis.com/billing/v1/invoices/${invoiceId}`,
            { custom_fields: customFields },
            {
                headers: {
                    'Authorization': `Zoho-oauthtoken ${accessToken}`,
                    'X-com-zoho-subscriptions-organizationid': orgId,
                    'Content-Type': 'application/json'
                }
            }
        );

        return response.data.invoice;
    } catch (error) {
        console.error('Error updating invoice:', error.response?.data || error.message);
        throw error;
    }
}

// Keep old subscription functions for backwards compatibility if needed
async function updateSubscriptionFields(subscriptionId, customFields) {
    const accessToken = await getZohoBillingAccessToken();
    const orgId = process.env.ZOHO_ORG_ID;

    try {
        console.log('Updating subscription', subscriptionId, 'with', customFields.length, 'custom fields');

        const response = await axios.put(
            `https://www.zohoapis.com/billing/v1/subscriptions/${subscriptionId}`,
            { custom_fields: customFields },
            {
                headers: {
                    'Authorization': `Zoho-oauthtoken ${accessToken}`,
                    'X-com-zoho-subscriptions-organizationid': orgId,
                    'Content-Type': 'application/json'
                }
            }
        );

        return response.data.subscription;
    } catch (error) {
        console.error('Error updating subscription:', error.response?.data || error.message);
        console.error('Payload was:', JSON.stringify({ custom_fields: customFields }, null, 2));
        throw error;
    }
}

// Format order (subscription with custom field address) for queue display
function formatOrderForQueue(record) {
    const shippingStatus = record.cf_shipping_status || '';
    const simCardNumbers = [];

    if (record.cf_sim_card_number) {
        simCardNumbers.push(record.cf_sim_card_number);
    }
    if (record.cf_secondary_sim_card_number) {
        simCardNumbers.push(record.cf_secondary_sim_card_number);
    }
    for (let i = 3; i <= 30; i += 1) {
        const fieldName = `cf_sim_card_number_${i}`;
        if (record[fieldName]) {
            simCardNumbers.push(record[fieldName]);
        }
    }

    // Use subscription custom fields for address
    const shippingAddress = {
        address: record.cf_street || '',
        address2: record.cf_street_2 || '',
        city: record.cf_city || '',
        state: record.cf_state || '',
        zip: record.cf_zip_code || '',
        country: record.cf_country || 'USA'
    };

    // Device type: use cf_device_type if set and not "Blank", otherwise "Sim Card Only"
    let deviceType = 'Sim Card Only';
    if (record.cf_device_type && record.cf_device_type !== 'Blank' && record.cf_device_type !== '') {
        deviceType = record.cf_device_type;
    }

    return {
        // Frontend expects invoice_id - use customer_id for mark-shipped lookup
        invoice_id: record.customer_id,
        invoice_number: record.subscription_number || record.name || '',
        // Show actual subscription ID and number
        subscription_id: record.subscription_id,
        subscription_number: record.subscription_number || record.name || '',
        customer_name: record.customer_name || '',
        email: record.email || '',
        phone: record.phone || record.mobile_phone || '',
        shipping_address: shippingAddress,
        // Device type from addon
        device_type: deviceType,
        status: shippingStatus === 'Shipped' ? 'shipped' : (record.cf_sim_card_number ? 'ready_to_ship' : 'pending_sim'),
        assigned_sim: record.cf_sim_card_number || '',
        sim_card_numbers: simCardNumbers,
        tracking_number: record.cf_tracking_number || '',
        line_status: record.cf_line_status || '',
        device_status: record.cf_device_status || '',
        ordered_by: record.cf_ordered_by || '',
        subscription_status: record.status || record.subscription_status || '',
        active_on_tmobile: record.cf_active_on_tmobile || '',
        tmobile_line_number: record.cf_tmobile_line_number || '',
        device_sn: record.cf_device_sn || record.cf_device_s_n || '',
        sim_card_quantity: record.cf_sim_card_quantity || '',
        device_quantity: record.cf_device_quantity || '',
        created_date: record.created_time,
        updated_date: record.updated_time,
        shipped_date: record.cf_shipping_date || null
    };
}

module.exports = {
    getZohoBillingAccessToken,
    getPendingOrders,
    getShippedOrders,
    getPendingTmobileActivations,
    getOrderDetails,
    updateInvoiceFields,  // Backwards compatibility (points to updateSubscriptionFields)
    updateSubscriptionFields,
    formatOrderForQueue,
    getCustomFieldValue
};

File: api/fulfillment/zoho_billing.py
"""
Zoho Billing API Integration Module
Handles OAuth2 authentication and API calls to Zoho Billing
"""

import os
import time
import requests
import logging
from datetime import datetime, timedelta
from typing import Dict, List, Optional

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class ZohoBillingAPI:
    """Zoho Billing API client with OAuth2 token management"""

    def __init__(self):
        self.client_id = os.getenv('ZOHO_CLIENT_ID')
        self.client_secret = os.getenv('ZOHO_CLIENT_SECRET')
        self.refresh_token = os.getenv('ZOHO_REFRESH_TOKEN')
        self.org_id = os.getenv('ZOHO_ORG_ID')
        self.base_url = 'https://www.zohoapis.com/billing/v1'

        # Token cache
        self._access_token = None
        self._token_expires_at = None

        if not all([self.client_id, self.client_secret, self.refresh_token, self.org_id]):
            logger.warning("Zoho credentials not fully configured in environment variables")

    def _get_access_token(self) -> str:
        """Get valid access token, refreshing if necessary"""

        # Return cached token if still valid
        if self._access_token and self._token_expires_at:
            if datetime.now() < self._token_expires_at - timedelta(minutes=5):
                return self._access_token

        # Refresh token
        logger.info("Refreshing Zoho access token")

        token_url = 'https://accounts.zoho.com/oauth/v2/token'
        params = {
            'refresh_token': self.refresh_token,
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'grant_type': 'refresh_token'
        }

        try:
            response = requests.post(token_url, params=params, timeout=10)
            response.raise_for_status()

            data = response.json()
            self._access_token = data['access_token']
            # Zoho tokens typically expire in 1 hour
            self._token_expires_at = datetime.now() + timedelta(seconds=data.get('expires_in', 3600))

            logger.info("Access token refreshed successfully")
            return self._access_token

        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to refresh access token: {e}")
            raise Exception(f"Zoho authentication failed: {e}")

    def _make_request(self, method: str, endpoint: str, params: Dict = None, json: Dict = None) -> Dict:
        """Make authenticated request to Zoho API"""

        access_token = self._get_access_token()
        url = f"{self.base_url}{endpoint}"

        headers = {
            'Authorization': f'Zoho-oauthtoken {access_token}',
            'X-com-zoho-subscriptions-organizationid': self.org_id,
            'Content-Type': 'application/json'
        }

        try:
            response = requests.request(
                method=method,
                url=url,
                headers=headers,
                params=params,
                json=json,
                timeout=30
            )

            response.raise_for_status()
            return response.json()

        except requests.exceptions.HTTPError as e:
            logger.error(f"Zoho API error: {e.response.status_code} - {e.response.text}")
            raise Exception(f"Zoho API error: {e.response.text}")
        except requests.exceptions.RequestException as e:
            logger.error(f"Request failed: {e}")
            raise Exception(f"Request to Zoho failed: {e}")

    # Core API Functions

    def get_pending_orders(self) -> List[Dict]:
        """
        Fetch invoices where Shipping Status = "Pending"
        Returns list of invoice objects
        """
        try:
            # Use custom field filter for Shipping Status
            params = {
                'filter_by': 'Status.All',
                'per_page': 100
            }

            response = self._make_request('GET', '/invoices', params=params)
            invoices = response.get('invoices', [])

            # Filter for pending shipping status
            pending = []
            for invoice in invoices:
                custom_fields = {cf['label']: cf.get('value', '')
                               for cf in invoice.get('custom_fields', [])}

                if custom_fields.get('Shipping Status') == 'Pending':
                    pending.append(invoice)

            logger.info(f"Found {len(pending)} pending orders")
            return pending

        except Exception as e:
            logger.error(f"Error fetching pending orders: {e}")
            raise

    def get_order_details(self, invoice_id: str) -> Dict:
        """
        Get full invoice details with customer data

        Args:
            invoice_id: Zoho invoice ID

        Returns:
            Complete invoice object with custom fields
        """
        try:
            response = self._make_request('GET', f'/invoices/{invoice_id}')
            invoice = response.get('invoice', {})

            logger.info(f"Fetched details for invoice {invoice_id}")
            return invoice

        except Exception as e:
            logger.error(f"Error fetching order {invoice_id}: {e}")
            raise

    def update_invoice_fields(self, invoice_id: str, fields: Dict) -> Dict:
        """
        Update custom fields on an invoice

        Args:
            invoice_id: Zoho invoice ID
            fields: Dict of field names to values, e.g.:
                    {
                        'SIM Card Number': '89011234567890123456',
                        'Tracking Number': '794644790327',
                        'Shipping Status': 'Shipped'
                    }

        Returns:
            Updated invoice object
        """
        try:
            # First get current invoice to preserve other fields
            current = self.get_order_details(invoice_id)

            # Build custom fields array
            custom_fields = current.get('custom_fields', [])
            custom_fields_dict = {cf['label']: cf for cf in custom_fields}

            # Update specified fields
            for field_name, field_value in fields.items():
                if field_name in custom_fields_dict:
                    custom_fields_dict[field_name]['value'] = field_value

            # Convert back to list
            updated_custom_fields = list(custom_fields_dict.values())

            # Update invoice
            payload = {
                'custom_fields': updated_custom_fields
            }

            response = self._make_request('PUT', f'/invoices/{invoice_id}', json=payload)

            logger.info(f"Updated invoice {invoice_id} fields: {list(fields.keys())}")
            return response.get('invoice', {})

        except Exception as e:
            logger.error(f"Error updating invoice {invoice_id}: {e}")
            raise

    def get_customer(self, customer_id: str) -> Dict:
        """
        Get customer details including shipping address

        Args:
            customer_id: Zoho customer ID

        Returns:
            Customer object with addresses
        """
        try:
            response = self._make_request('GET', f'/customers/{customer_id}')
            customer = response.get('customer', {})

            logger.info(f"Fetched customer {customer_id}")
            return customer

        except Exception as e:
            logger.error(f"Error fetching customer {customer_id}: {e}")
            raise

    def list_invoices_by_status(self, status: str = 'all') -> List[Dict]:
        """
        List invoices filtered by status

        Args:
            status: 'all', 'sent', 'draft', 'overdue', 'paid', 'void', 'unpaid'

        Returns:
            List of invoice objects
        """
        try:
            status_map = {
                'all': 'Status.All',
                'sent': 'Status.Sent',
                'draft': 'Status.Draft',
                'overdue': 'Status.Overdue',
                'paid': 'Status.Paid',
                'void': 'Status.Void',
                'unpaid': 'Status.Unpaid'
            }

            params = {
                'filter_by': status_map.get(status, 'Status.All'),
                'per_page': 100
            }

            response = self._make_request('GET', '/invoices', params=params)
            invoices = response.get('invoices', [])

            logger.info(f"Fetched {len(invoices)} invoices with status '{status}'")
            return invoices

        except Exception as e:
            logger.error(f"Error listing invoices: {e}")
            raise

    def get_custom_field_value(self, invoice: Dict, field_name: str) -> Optional[str]:
        """
        Helper to extract custom field value from invoice

        Args:
            invoice: Invoice object
            field_name: Name of custom field

        Returns:
            Field value or None if not found
        """
        custom_fields = invoice.get('custom_fields', [])
        for field in custom_fields:
            if field.get('label') == field_name:
                return field.get('value')
        return None


# Singleton instance
_zoho_api = None

def get_zoho_api() -> ZohoBillingAPI:
    """Get or create Zoho API singleton instance"""
    global _zoho_api
    if _zoho_api is None:
        _zoho_api = ZohoBillingAPI()
    return _zoho_api
